### PR TITLE
ux(Stencil): Enhance UX for stencils

### PR DESCRIPTION
**Pull request by ccailly**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 

### Summary

This pull request is a continuation of the UX improvements introduced in PR #15229, with a focus on enhancing the Stencil Editor by implementing AJAX for the "Add" and "Remove" buttons.

### Changes

- Added keyboard shortcuts for faster interaction.
- Integrated AJAX for "Add" and "Remove" actions to improve user experience.
- Limited input values to prevent browser crashes.
- Updated code structure for the above changes.
- Set a maximum zone number to 128 for consistency.
- Validated input for the zone number.

### CODE REVIEW

**Review by orthagh — COMMENTED** [review]
A quick review for some misses:
- [Enter] and [Esc] key only work when focusing editor. when you focus the form (editing the port label for example), the don't work.
- "Set a maximum zone number to 128 for consistency.". It should be rare, but i found some multi-U enclosures (not a direct equipment so) with 128 front ports and a few back ports (console). I suggest to increase a bit (the double ?)

File: js/stencil-editor.js [hunk 1]
```diff
@@ -34,12 +34,21 @@
...
+ */
 const StencilEditor = function (container, rand, zones_definition) {
-    let zones = zones_definition;
+    var zones = zones_definition;
 
-    let croppers = [];
+    var croppers = [];
 
-    let _this = this;
+    var _this = this;
```
Comment (AdrienClairembault) [inline js/stencil-editor.js:51]: Why did you need to use `var` instead of `let` here ?

**Review by AdrienClairembault — CHANGES_REQUESTED** [review]
Looks good to me👍

**Comment by AdrienClairembault** [comment]
Another small UX annoyance I've found is that the `+` and  `-` buttons move each time you click on them (as doing so add/remove item from the list and the button are a the end of the list).
So if you want to add or delete multiple items you have to keep moving your mouse each time.

Maybe we could move them on the line above (next to "Define port data in image") ?

**Comment by AdrienClairembault** [comment]
>     * [Enter] and [Esc] key only work when focusing editor. when you focus the form (editing the port label for example), the don't work.

Weird, this part works fine on my side (firefox).

**Review by AdrienClairembault — APPROVED** [review]

File: js/stencil-editor.js [hunk 2]
```diff
@@ -105,18 +120,39 @@ const StencilEditor = function (container, rand, zones_definition) {
             });
 
         // keyboard events
-        $('#zone_label-' + rand)
-            .on('keypress', function (e) {
+        $(document)
+            .on('keyup', function (e) {
```
Comment (cedric-anne) [inline js/stencil-editor.js:124]: Listening keyup/keypress events and prevent their default behaviour on the whole document may result in unexpected side effets, for instance for people that uses the keyboard navigation.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: js/stencil-editor.js (hunk 2 above)
Comment (ccailly) [reply js/stencil-editor.js:124]: I've tried to restrict listening only to the container containing the stencil, but sometimes I get strange behavior: I get the impression that sometimes the container loses focus, which has the effect of breaking key shortcuts.

I've fine-tuned shortcut detection so that they are only triggered when the editor is active. So, for example, if a user wants to perform an "escape" and the editor is active, the shortcut will only be caught once by the stencil. If the user wanted to "escape" to another action, he'd just have to do it a second time.

I think this solution is acceptable, as I haven't managed to restrict listening to the container only.

**Comment by ccailly** [comment]
> A quick review for some misses:
> 
>     * [Enter] and [Esc] key only work when focusing editor. when you focus the form (editing the port label for example), the don't work.
> 
>     * "Set a maximum zone number to 128 for consistency.". It should be rare, but i found some multi-U enclosures (not a direct equipment so) with 128 front ports and a few back ports (console). I suggest to increase a bit (the double ?)

After numerous tests in Firefox and Chrome, I can't reproduce the bug. The shortcuts still work, even when I'm focused in the form inputs.
Can you tell me more?

**Review by AdrienClairembault — APPROVED** [review]

File: js/stencil-editor.js [hunk 3]
```diff
@@ -105,18 +120,41 @@ const StencilEditor = function (container, rand, zones_definition) {
...
-                    _this.saveZoneData();
+
+                // Check if one of the cropper as a selection
+                const hasSelection = croppers.some(function (cropper) {
+                    return cropper.getCropperSelection() !== undefined
+                        && cropper.getCropperSelection().height > 0
+                        && cropper.getCropperSelection().width > 0;
+                });
+
+                if (_this.isEditorActive()) {
```
Comment (cedric-anne) [inline js/stencil-editor.js:134]: Could be moved at tthe top of the function, to prevent useless calculations:

```js
if (!_this.isEditorActive()) {
    return;
}
```

File: js/stencil-editor.js [hunk 4]
```diff
@@ -105,18 +120,41 @@ const StencilEditor = function (container, rand, zones_definition) {
...
+                    } else if (keycode == 27) {
+                        _this.editorDisable();
+                        e.preventDefault();
+                    }
                 }
-            });
-        $('#zone_number-' + rand)
+            })
             .on('keypress', function (e) {
                 var keycode = (e.keyCode ? e.keyCode : e.which);
```
Comment (cedric-anne) [inline js/stencil-editor.js:145]:
```suggestion
                if (!_this.isEditorActive()) {
                    return;
                }

                var keycode = (e.keyCode ? e.keyCode : e.which);
```

File: templates/stencil/editor.html.twig [hunk 5]
```diff
@@ -106,15 +113,15 @@
                     </div>
                 </div>
                 <div class="col-auto">
-                    <button type="button" id="save-zone-data-{{ rand }}" class="btn btn-primary">
+                    <button type="button" id="save-zone-data-{{ rand }}" class="btn btn-primary" data-bs-toggle="tooltip" data-bs-placement="right" data-bs-title="{{ params['save_zone_data_label']|default(__("Save zone data")) ~ " (Enter)" }}">
```
Comment (cedric-anne) [inline templates/stencil/editor.html.twig:116]: 1. `Enter` should be translated.
2. As translation may contains unexpected chars, it is safer to use the `html_attr` escape filter.
```suggestion
                    <button type="button" id="save-zone-data-{{ rand }}" class="btn btn-primary" data-bs-toggle="tooltip" data-bs-placement="right" data-bs-title="{{ (params['save_zone_data_label']|default(__("Save zone data")) ~ ' (' ~ _x('keyboard', 'Enter') ~ ')')|escape('html_attr') }}">
```

File: templates/stencil/editor.html.twig [hunk 6]
```diff
@@ -106,15 +113,15 @@
...
                         <i class="ti ti-device-floppy"></i>
                         {% if params.save_zone_data_label is defined %}
                             <span>{{ params['save_zone_data_label'] }}</span>
                         {% else %}
                             <span>{{ __("Save zone data") }}</span>
                         {% endif %}
                     </button>
-                    <button type="button" id="cancel-zone-data-{{ rand }}" class="btn btn-outline-danger">
+                    <button type="button" id="cancel-zone-data-{{ rand }}" class="btn btn-outline-danger" data-bs-toggle="tooltip" data-bs-placement="right" data-bs-title="{{ __("Cancel") ~ " (Esc)" }}">
```
Comment (cedric-anne) [inline templates/stencil/editor.html.twig:124]:
```suggestion
                    <button type="button" id="cancel-zone-data-{{ rand }}" class="btn btn-outline-danger" data-bs-toggle="tooltip" data-bs-placement="right" data-bs-title="{{ (__("Cancel") ~ ' (' ~ _x('keyboard', 'Esc') ~ ')')|escape('html_attr') }}">
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Comment by orthagh** [comment]
After checking together, the issue with keys shortcuts arrives only when no data yet saved. So it doesn't matter.
Another thing, we cannot clear a port data

**Review by orthagh — APPROVED** [review]
